### PR TITLE
2D fillet chamfer

### LINKED ISSUE #348: Expose 2D fillet and 2D chamfer

### PR DESCRIPTION
**Pull request by adam-urbanczyk**
Will resolve #348

### CODE REVIEW

**Comment by adam-urbanczyk** [comment]
Open question: do we want to expose those functions in the fluent API?

**Review by jmwright — APPROVED** [review]
Thanks @adam-urbanczyk

File: tests/test_cadquery.py [hunk 1]
```diff
@@ -4269,3 +4269,23 @@ def testCompSolid(self):
...
+    def test2Dfillet(self):
+
+        r = Workplane().rect(1, 2).wires().val()
+        f = Face.makeFromWires(r)
+        verts = r.Vertices()
+
+        self.assertEqual(len(f.fillet2D(0.5, verts).Vertices()), 6)
+        self.assertEqual(len(r.fillet2D(0.5, verts).Vertices()), 6)
+        self.assertEqual(len(r.fillet2D(0.25, verts).Vertices()), 8)
+
```
Comment (marcus7070) [inline tests/test_cadquery.py:4282]:
```suggestion
        # Test fillet2D with open wire and single vertex
        w0 = Workplane().hLine(1).vLine(1).wire()
        w0_verts = w0.vertices(">X and <Y").vals()
        unfilleted_wire0 = w0.val()
        filleted_wire0 = unfilleted_wire0.fillet2D(0.5, w0_verts)
        # the filleted wire is shorter than the original
        self.assertGreater(unfilleted_wire0.Length() - filleted_wire0.Length(), 0.1)

```

File: cadquery/occ_impl/shapes.py [hunk 2]
```diff
@@ -1974,6 +2019,35 @@ def makeFromWires(
...
+
+        chamfer_builder = BRepFilletAPI_MakeFillet2d(self.wrapped)
+        edge_map = self._entitiesFrom("Vertex", "Edge")
+
+        for e1, e2 in edge_map.values():
+            chamfer_builder.AddChamfer(TopoDS.Edge_s(e1), TopoDS.Edge_s(e2), d, d)
+
+        chamfer_builder.Build()
+
+        return self.__class__(chamfer_builder.Shape()).fix()
```
Comment (marcus7070) [inline cadquery/occ_impl/shapes.py:2049]: `chamfer2D` does not use the `vertices` argument?

It also errors with an open wire, like 
```python
w0 = cq.Workplane().hLine(1).vLine(1).wire().val().chamfer2D(0.2, [])
```
```
>       for e1, e2 in edge_map.values():
E       ValueError: not enough values to unpack (expected 2, got 1)

cadquery/occ_impl/shapes.py:2044: ValueError
```

**Review by marcus7070 — CHANGES_REQUESTED** [review]
Sorry, the online OCCT docs are down right now, otherwise I would help out more with that chamfer2D stuff.

File: cadquery/occ_impl/shapes.py (hunk 2 above)
Comment (adam-urbanczyk) [reply cadquery/occ_impl/shapes.py]: Should be resolved now. Thanks for spotting this!

**Review by marcus7070 — APPROVED** [review]

**Comment by marcus7070** [comment]
Thanks @adam-urbanczyk, this stuff looks great! That `Shape._entitiesFrom` method looks very powerful, I hope we get to do some other things with it!

> Open question: do we want to expose those functions in the fluent API?

Yeah, that would be great. I can think of a few options on how we do it in a fluent way, I'll start a new issue to discuss.

**Pull request merged**